pump/storage/storage.go

```diff
@@ -519,34 +520,67 @@ func (a *Append) GCTS(ts int64) {
 	atomic.StoreInt64(&a.gcTS, ts)
 	a.saveGCTSToDB(ts)
 	gcTSGauge.Set(float64(oracle.ExtractPhysical(uint64(ts))))
+
+	if atomic.LoadInt32(&a.gcWorking) == 1 {
+		return
+	}
+
 	// for commit binlog TS ts_c, we may need to get the according P binlog ts_p(ts_p < ts_c
 	// so we forward a little bit to make sure we can get the according P binlog
-	a.doGCTS(ts - int64(oracle.EncodeTSO(maxTxnTimeoutSecond*1000)))
+	go a.doGCTS(ts - int64(oracle.EncodeTSO(maxTxnTimeoutSecond*1000)))
 }
 
 func (a *Append) doGCTS(ts int64) {
 	log.Info("start gc ts: ", ts)
-	irange := &util.Range{
-		Start: encodeTSKey(0),
-		Limit: encodeTSKey(ts + 1),
-	}
-	iter := a.metadata.NewIterator(irange, nil)
+
+	atomic.StoreInt32(&a.gcWorking, 1)
+	defer atomic.StoreInt32(&a.gcWorking, 0)
+
 	batch := new(leveldb.Batch)
-	for iter.Next() {
-		batch.Delete(iter.Key())
-		if batch.Len() == 1024 {
-			err := a.metadata.Write(batch, nil)
-			if err != nil {
-				log.Error(err)
-			}
-			batch.Reset()
-		}
+	l0Trigger := defaultStorageKVConfig.CompactionL0Trigger
+	if a.options.KVConfig != nil && a.options.KVConfig.CompactionL0Trigger > 0 {
+		l0Trigger = a.options.KVConfig.CompactionL0Trigger
 	}
 
-	if batch.Len() > 0 {
-		err := a.metadata.Write(batch, nil)
+	for {
+		var stats leveldb.DBStats
+		err := a.metadata.Stats(&stats)
 		if err != nil {
 			log.Error(err)
+			time.Sleep(5 * time.Second)
+			continue
+		}
+		if stats.LevelTablesCounts[0] >= l0Trigger {
```

Comment:
output log

Reply:
add in https://github.com/pingcap/tidb-binlog/pull/620/commits/a82e66fc7c7ec709168ea5211a82cc9cf0de5217